text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,293 @@
+
+```rust
+// `array_vec` can store up to 64 elements
+let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+```
+
+Of course, fixed buffers lead to inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+```rust
+// This vector can store up to 0 elements, therefore, nothing at all
+let mut array_vec: ArrayVec<i32, 0> = ArrayVec::new();
+let push_result = array_vec.push(1);
+// Ooppss... Our push operation wasn't successful
+assert!(push_result.is_err());
+```
+
+A good question is: Should I use `core::collections::ArrayVec<T>` or `alloc::collections::Vec<T>`? Well, `Vec` is already good enough for most situations while stack allocation usually shines for small sizes.
+
+* Do you have a known upper bound?
+
+* How much memory are you going to allocate for your program? The default values of `RUST_MIN_STACK` or `ulimit -s` might not be enough.
+
+* Are you using nested `Vec`s? `Vec<ArrayVec<T, N>>` might be better than `Vec<Vec<T>>`.
+
+Each use-case is different and should be pondered individually. In case of doubt, stick with `Vec`.
+
+For a more technical overview, take a look at the following operations:
+
+```rust
+// `array_vec` has a pre-allocated memory of 2048 bits (32 * 64) that can store up
+// to 64 decimals.
+let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+
+// Although reserved, there isn't anything explicitly stored yet
+assert_eq!(array_vec.len(), 0);
+
+// Initializes the first 32 bits with a simple '1' decimal or
+// 00000000 00000000 00000000 00000001 bits
+array_vec.push(1);
+
+// Our vector memory is now split into a 32/2016 pair of initialized and
+// uninitialized memory respectively
+assert_eq!(array_vec.len(), 1);
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`ArrayVec` is a contiguous memory block where elements can be collected, therefore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+
+The API basically mimics most of the current `Vec` surface with some tweaks to manage capacity.
+
+Notably, these tweaked methods are fallible versions of some well-known functions like `push` that will return `Result` instead of panicking at run-time. Since the upper capacity bound is known at compile-time, the compiler is likely going to remove most of the conditional bounding checking.
+
+```rust
+// Please, bare in mind that these methods are simply suggestions. Discussions about the
+// API should probably take place elsewhere.
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
+    len: usize,
```

Comment:
Should it be usize when we probably can't even hit u16?

Reply:
"Probably" is not a great fit for the standard library.

Ideally we’d use specialization of some private trait with an associated type to choose the smallest integer type that can represent `0..=N`. However it looks like boolean conditions based on `const` parameters can not be used in `where` clauses today, and I don’t know if supporting that is planned at all. And this would only help where `align_of::<T>()` is less than `size_of::<usize>()`.

Reply:
Potential hack until specializations works: Store the size in a `[u8; M]` which gets converted to a usize on demand. Calculate the `M` based on `N`.